applications/conceptual-model-editor/src/diagram/diagram-controller.ts

```diff
@@ -216,6 +258,45 @@ const createChangeSelectionHandler = () => {
     // As a result the toolbar was open and closed, causing a blink.
     // The solution of choice was to draw an inspiration from NodeToolbar
     // and watch for edge selection in EdgeToolbar.
+
+    setSelectedNodes(prevSelectedNodes => {
+      const diagramNodes = reactFlow.getNodes();
+      const newlySelected = [];
+      const newlyRemovedFromSelection = [];
+
+      for (const node of diagramNodes) {
+        const wasSelected = prevSelectedNodes.find(selectedNodeIdentifier => selectedNodeIdentifier === node.id) !== undefined;
+        if(wasSelected) {
+          const sameNodeFromSelection = nodes.find(nnode => node.id === nnode.id);
+          const isNoLongerSelected = sameNodeFromSelection === undefined;
+          if(isNoLongerSelected) {
+            newlyRemovedFromSelection.push(node);
+          }
+        }
+        else {
+          const sameNodeFromSelection = nodes.find(nnode => node.id === nnode.id);
+          const isNewlySelected = sameNodeFromSelection !== undefined;
+          if(isNewlySelected) {
+            newlySelected.push(node);
+          }
+        }
+      }
+
+      const newSelectedNodes = nodes.map((node) => node.id);
+      let insertPosition = 0;
+      for(let i = 0; i < prevSelectedNodes.length; i++) {
```

Comment:
Basically the idea here and of the whole method is that reactflow doesn't keep the order of selection, which we need to show the selection toolbar on top of the lastly selected node. So we just go through the previous selection and those nodes which are also in the new selection are put to correct position in the if few lines down from here.

So that it is the reason why need this method and explicitly remember the `selectedNodes` and `selectedEdges`.

But looking at the code I have no idea why do I even perform the previous `for .. of`, since I dont do anything with the arrays, I guess that I was just going for different implementation and then for some reason switched.